### PR TITLE
Adding procedural test generation to test interactions between various operators.

### PR DESCRIPTION
**Pull request by maumar**
Testing up to 6 sources (3 pairs for predicate, 2 trios for projection). Entities have 3-4 max rows so even if all 6 are used the result set is not too large. Unused sources are pruned so usually we end up using 3-4 sources.

### CODE REVIEW

File: test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs [hunk 1]
```diff
@@ -0,0 +1,1784 @@
...
+    protected ExpectedQueryRewritingVisitor ExpectedQueryRewriter { get; init; }
+
+    private static readonly MethodInfo _likeMethodInfo
+        = typeof(DbFunctionsExtensions).GetRuntimeMethod(
+            nameof(DbFunctionsExtensions.Like), new[] { typeof(DbFunctions), typeof(string), typeof(string) });
+
+    private static readonly MethodInfo _stringConcatMethodInfo
+        = typeof(string).GetRuntimeMethod(
+            nameof(string.Concat), new[] { typeof(string), typeof(string) });
+
```
Comment (roji) [inline test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs:24]: empty line

File: test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs [hunk 2]
```diff
@@ -0,0 +1,1784 @@
...
+public abstract class OperatorsQueryTestBase : NonSharedModelTestBase
+{
+    protected readonly List<((Type, Type) InputTypes, Type ResultType, Func<Expression, Expression, Expression> OperatorCreator)> Binaries;
+    protected readonly List<(Type InputType, Type ResultType, Func<Expression, Expression> OperatorCreator)> Unaries;
+    protected readonly Dictionary<Type, Type> PropertyTypeToEntityMap;
+
+    protected OperatorsData ExpectedData { get; init; }
+    protected ExpectedQueryRewritingVisitor ExpectedQueryRewriter { get; init; }
+
+    private static readonly MethodInfo _likeMethodInfo
```
Comment (roji) [inline test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs:17]:
```suggestion
    private static readonly MethodInfo LikeMethodInfo
```

And below

File: test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs [hunk 3]
```diff
@@ -0,0 +1,1784 @@
...
+            ((typeof(bool), typeof(bool)), typeof(bool), Expression.NotEqual),
+            ((typeof(bool), typeof(bool)), typeof(bool), Expression.AndAlso),
+            ((typeof(bool), typeof(bool)), typeof(bool), Expression.OrElse),
+            ((typeof(bool), typeof(bool)), typeof(bool), Expression.And),
+            ((typeof(bool), typeof(bool)), typeof(bool), Expression.Or),
+        };
+
+        Unaries = new()
+        {
+            (typeof(string), typeof(bool), x => Expression.Equal(x, Expression.Constant(null, typeof(string)))),
```
Comment (roji) [inline test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs:83]: I think unary equality (as equality to null) is our own thing in EF (right?); in other words, are we expected to actually translate this in a query expression tree given by the user? If it's only an internal temporary representation thing, we probably don't need to be testing for it here?

File: test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs [hunk 4]
```diff
@@ -0,0 +1,1784 @@
...
+            Assert.Equal(expected[i], actual[i]);
+        }
+    }
+
+    //[ConditionalFact]
+    public virtual async Task Procedural_predicate_test_six_sources_three_pairs()
+    {
+        var maxDepth = 7;
+        var contextFactory = await InitializeAsync<OperatorsContext>(seed: Seed);
+        using (var context = contextFactory.CreateContext())
```
Comment (roji) [inline test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs:262]: using statement (no block needed)

File: test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs [hunk 5]
```diff
@@ -0,0 +1,1784 @@
...
+            var actualSetSource = new ActualSetSource(context);
+
+            while (true)
+            {
+                var seed = new Random().Next();
+                var random = new Random(seed);
+
+                var possibleTypes = OperatorsData.Instance.ConstantExpressionsPerType.Keys.ToArray();
+
+                var typesUsed = new bool[6];
```
Comment (roji) [inline test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs:273]: unused

File: test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs [hunk 6]
```diff
@@ -0,0 +1,1784 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+
+namespace Microsoft.EntityFrameworkCore.Query;
+
+public abstract class OperatorsQueryTestBase : NonSharedModelTestBase
+{
+    protected readonly List<((Type, Type) InputTypes, Type ResultType, Func<Expression, Expression, Expression> OperatorCreator)> Binaries;
```
Comment (roji) [inline test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs:10]:
```suggestion
    protected readonly List<((Type Left, Type Right) InputTypes, Type ResultType, Func<Expression, Expression, Expression> OperatorCreator)> Binaries;
```

File: test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs [hunk 7]
```diff
@@ -0,0 +1,1784 @@
...
+                    currentDepth,
+                    maxDepth,
+                    operation,
+                    types,
+                    rootPropertyExpressions,
+                    possibleBinaries,
+                    possibleUnaries);
+            }
+        }
+        else
```
Comment (roji) [inline test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs:451]: Can remove

File: test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs [hunk 8]
```diff
@@ -0,0 +1,1784 @@
...
+        Type[] types,
+        RootEntityExpressionInfo[] rootPropertyExpressions,
+        List<((Type, Type) InputTypes, Type ResultType, Func<Expression, Expression, Expression> OperatorCreator)> possibleBinaries,
+        List<(Type InputType, Type ResultType, Func<Expression, Expression> OperatorCreator)> possibleUnaries)
+    {
+        // see if we want additional level of nesting, the deeper we go the lower the probability
+        // we also force nesting if we end up with an expected node that we don't have the root entity for
+        // this can happen when we use convert - e.g. we only have int sources, but we expect long
+        var rollAddDepth = random.Next(maxDepth);
+        if (rollAddDepth >= currentDepth)
```
Comment (roji) [inline test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs:412]: I'm not sure I'm understanding this correctly, but this seems to nest multiple expressions up to this depth... If so, I think that for precedence issues it's always enough to check just two expressions (one nested within the other) - is that a possible simplification here? Or am I totally misunderstanding?

File: test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs [hunk 9]
```diff
@@ -0,0 +1,1784 @@
...
+        {
+            var constants = OperatorsData.Instance.ConstantExpressionsPerType[currentResultType];
+
+            return constants[random.Next(constants.Count)];
+        }
+
+        var unusedExpressions = matchingExpressions.Where(x => !x.Used).ToList();
+        if (unusedExpressions.Any())
+        {
+            var chosenExpresion = unusedExpressions[random.Next(unusedExpressions.Count)];
```
Comment (roji) [inline test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs:478]:
```suggestion
            var chosenExpression = unusedExpressions[random.Next(unusedExpressions.Count)];
```

File: test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs [hunk 10]
```diff
@@ -0,0 +1,1784 @@
...
+            chosenExpresion.Used = true;
+
+            return chosenExpresion.Expression;
+        }
+        else
```
Comment (roji) [inline test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs:483]: can remove (also in other places in this file)

**Review by roji — APPROVED** [review]
This is... complex :rofl: 

But :sheep: :it:, most of this is manual testing only - and if it caught some bugs, it's great.

**Comment by roji** [comment]
BTW we have various existing tests which check parentheses stuff, these can in theory be moved here at some point, so we concentrate all the operator/parentheses tests in one place. Not necessarily worth spending time on now, but it's good to have the this one place now.

**Comment by roji** [comment]
One last thought - because of the complexity of the procedural testing, it may be worth splitting the (complex) procedural testing to something like OperatorsProceduralQueryTestBase, so that we have the non-procedural regression tests (Regression_test1, 2...) in one easy, clear place.

File: test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs [hunk 11]
```diff
@@ -0,0 +1,1784 @@
...
+        ctx.Set<OperatorEntityLong>().AddRange(ExpectedData.OperatorEntitiesLong);
+        ctx.Set<OperatorEntityBool>().AddRange(ExpectedData.OperatorEntitiesBool);
+        ctx.Set<OperatorEntityNullableBool>().AddRange(ExpectedData.OperatorEntitiesNullableBool);
+        ctx.Set<OperatorEntityDateTimeOffset>().AddRange(ExpectedData.OperatorEntitiesDateTimeOffset);
+
+        ctx.SaveChanges();
+    }
+
+    [ConditionalFact(Skip = "issue #30245")]
+    public virtual async Task Regression_test1()
```
Comment (roji) [inline test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs:147]: Consider giving these full names which describe what is tested

Comment (maumar) [reply test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs:147]: I can see a problem with that, depending on complexity of the query :D But it sounds good for the simple stuff we have now

File: test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs (hunk 3 above)
Comment (maumar) [reply test/EFCore.Relational.Specification.Tests/Query/OperatorsQueryTestBase.cs:83]: We use `ExpressionType.Equal`/`ExpressionType.Equal` internally in EF to represent null equality, but here there reason is a bit different. Those Unaries/Binaries lists store operations you can perform given the input, and it tells you the type of output. So here we have input being string (however we arrived at that input type), we know we can apply `== null` operation and we get bool back. So, say, we have two source entities (a, b) that have string values, as first step we may have generated string concat operation: a.Value + b.Value (two string inputs, and string output), and then in second step, we have string input (i.e. the result of concat), and we can apply the == null. So the end result would be

from a in ctx.Set<OperatorEntityString>
from b in ctx.Set<OperatorEntityString>
select (a.Value + b.Value) == null

which is somewhat interesting from parens perspective:

`(a.Value + b.Value) IS NULL` vs `a.Value + (b.Value IS NULL)`

In other words, in this list just look at the lambda, which is the actual expression that will get generated (and is a normal binary), but stored it in Unary, because it only really has one variable argument, the second one being null. Same goes for arg LIKE 'A%' - technically LIke is a binary operation, but the second argument is fixed so we store as unary here
